flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/PreciseTimestamp.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.dataformat;
+
+import org.apache.flink.table.types.logical.TimestampType;
+import org.apache.flink.util.Preconditions;
+
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+
+import static org.apache.calcite.avatica.util.DateTimeUtils.timestampStringToUnixDate;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestamp;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestampToString;
+
+/**
+ * Immutable SQL TIMESTAMP type with nanosecond precision.
+ */
+public class PreciseTimestamp implements Comparable<PreciseTimestamp> {
+
+	public static final int MAX_PRECISION = TimestampType.MAX_PRECISION;
+
+	public static final int MIN_PRECISION = TimestampType.MIN_PRECISION;
+
+	private final long milliseconds;
+
+	private final int nanoseconds;
+
+	private final int precision;
+
+	private PreciseTimestamp(long milliseconds, int nanoseconds, int precision) {
+		Preconditions.checkArgument(precision >= MIN_PRECISION && precision <= MAX_PRECISION);
+		Preconditions.checkArgument(nanoseconds >= 0 && nanoseconds <= 999_999_999);
+		this.milliseconds = milliseconds;
+		this.nanoseconds = nanoseconds;
+		this.precision = precision;
+	}
+
+	@Override
+	public int compareTo(PreciseTimestamp that) {
+		return this.milliseconds == that.milliseconds ?
+			Integer.compare(this.nanoseconds, that.nanoseconds) :
+			Long.compare(this.milliseconds, that.milliseconds);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (!(obj instanceof PreciseTimestamp)) {
+			return false;
+		}
+		PreciseTimestamp that = (PreciseTimestamp) obj;
+		return this.compareTo(that) == 0;
+	}
+
+	@Override
+	public String toString() {
```

Comment:
You can just use `toLocalDateTime().format(DateTimeFormatter)`.

Reply:
Just use `toLocalDateTime().toString()` now. 
My original thought is obtain different string representation according to precision.
Since we drop precision in this class, `LocalDateTime::toString` is good now.